### DISCUSSION TITLE
We should consider switching the frontend codebase to TypeScript

### DISCUSSION THREAD

**Discussion by VIKTORVAV99** [electricitymaps/electricitymaps-contrib#4077] [Ideas]
# We should consider switching the frontend codebase to TypeScript.

### Benefits
* It would allow us to type check and validate the code at build time instead of runtime.
* It would allow us to clearly define options such as the valid iconNames in #3996.
* Together with a more robust CI pipeline we would get higher confidence in dependency updates and code changes.
* We can change code incrementally starting with smaller components that rely on passing the correct string values such as the `<Icon/>` component in #3996 where it would see the most immediate benefits.

### Drawbacks
* It would require substantial changes to the codebase and would likely take a lot of time to switch everything to TypeScript.
* It would bring more complexity to the codebase.
* It will increase the complexity for webpack, babel and eslint (plugins and presets makes this rather simple but still)

If anyone have any other benefits/drawbacks or just want to discuss the changes needed feel free to comment so we can plan the move forward no matter the direction.

**Comment by madsnedergaard** [comment]
Thanks for bringing this up! We've talked about it before internally, and one additional drawback not mentioned here is "approachability" aka. this would make it harder for developers that don't know TS to contribute to the code. 
Except for that I very much agree with the points you're mentioning :)

**Reply by VIKTORVAV99** [reply]
I agree it would be harder for first time contributors and developers who haven't really used typescript before, but I believe the the benefits far outweigh the downsides of this long term.

Perhaps you can bring it up internally again and if it's something you want to go forward with it we can try it out on a few limited components and go from there?
